src/optimizing/Analyze_rule.ml

```diff
@@ -205,13 +205,20 @@ let rec (cnf : Rule.formula -> cnf_step0) =
       let zs = Common.map (fun (_t, cond) -> And [ Or [ LCond cond ] ]) conds in
       And (ys @ zs |> List.concat_map (function And ors -> ors))
   | R.Or (_, xs) ->
-      let is_dangerously_large l = List.compare_length_with l 1_000_000 > 0 in
+      let is_dangerously_large p q =
+        List.compare_length_with p 1_000_000 > 0
+        || List.compare_length_with q 1_000_000 > 0
+        ||
+        let p_len = List.length p in
```

Comment:
Once we use `List.length` here then it would be better to refactor the code above to just use `p_len` and `q_len`, it becomes pointless to use `List.compare_length_with`.

Reply:
Actually I'm not sure we even need `List.compare_length_with p 1_000_000 > 0  || List.compare_length_with q 1_000_000 > 0`, probably the new condition suffices.

Reply:
OK I saw your PR comment, my bad not having read it before, although I think that comment belonged here in the code.

Reply:
I think using `List.compare_length_with` here was an acceptable "premature optimization" because it didn't really add any complexity and it could potentially speed up things. But computing the length of a large list should be pretty fast. And the problem with how it looks now is that the _normal_ path will be to evaluate the three Boolean conditions (because the CNF will rarely explode), so we're just making this check (a bit) slower by keeping the previous conditions. And the new condition is _almost_ stricter than the previous ones anyways. You could have `ps` having 1_000_001 elements, and `qs` having up to 9 elements without your new condition becoming true but... does it matter? If it matters then better to just change your condition to `p_len > Int.div 1_000_000 q_len`.

Reply:
That's a fair objection. I figured List.length was fast enough it didn't matter, and there is the slight chance that a list could be truly pathologically long. But I wasn't sure if that could actually happen without the rule being somehow invalid. That's why I made the comment a PR comment, I expected that I would change the code.

I don't really have strong feelings, the numbers here are largely based on vibes. (I know that 50 million and 100 million are definitely too long but not what the lower bound is)